data-resource/src/crc32.rs

```diff
@@ -0,0 +1,75 @@
+use crate::ResourceIdTrait;
+use data_error::Result;
+use serde::{Deserialize, Serialize};
+use std::{
+    fs,
+    io::{BufRead, BufReader},
+    path::Path,
+};
+
+use crc32fast::Hasher;
+
+/// Represents a resource identifier using the CRC32 algorithm.
+///
+/// Uses [`crc32fast`] crate to compute the hash value.
+#[derive(
+    Eq,
+    Ord,
+    PartialEq,
+    PartialOrd,
+    Hash,
+    Clone,
+    Copy,
+    Debug,
+    Deserialize,
+    Serialize,
+)]
+pub struct ResourceId {}
```

Comment:
It was like this:
```
pub struct ResourceId {
    pub data_size: u64,
    pub hash: u32,
}
```

By the way, could it be useful to have definition like this shared across implementations?
```
pub struct ResourceId<H> {
    pub data_size: u64,
    pub hash: H,
}
```

Then we would have `ResourceId` type everywhere in the client code without the need for aliases.

Reply:
We can use a generic type instead of an associated type
```rust
pub trait ResourceIdTrait<T>
where
    T: Debug
        + Display
        + FromStr
        + Clone
        + PartialEq
        + Eq
        + Ord
        + PartialOrd
        + Hash
        + Serialize,
{
    /// Computes the resource identifier from the given file path
    fn from_path<P: AsRef<Path>>(file_path: P) -> Result<T>;
    /// Computes the resource identifier from the given bytes
    fn from_bytes(data: &[u8]) -> Result<T>;
}
```
and in `data-resource/src/crc32.rs`, we can implement it 
```rust
impl ResourceIdTrait<u32> for ResourceId {
    fn from_path<P: AsRef<Path>>(file_path: P) -> Result<u32> {
```
I'm not entirely sure how that change would affect our situation.

Btw, do you think we should keep `data_size` as a field of `ResourceId`?

Reply:
Another point to keep in mind:
With `pub trait ResourceIdTrait<T> where T: Debug`, we'd need to manually annotate types when importing `ResourceId` in other crates. 

This isn't ideal as we want to switch between "non-cryptographic-hash" and "cryptographic-hash" implementations without manual code changes if possible.